test/unit/reductions/exponentiated_gradient/test_lagrangian.py

```diff
@@ -0,0 +1,110 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+import pandas as pd
+import pickle
+import pytest
+
+from fairlearn.reductions._exponentiated_gradient._lagrangian import _Lagrangian
+from fairlearn.reductions import DemographicParity, EqualizedOdds
+
+from .test_utilities import _get_data
+from .simple_learners import LeastSquaresBinaryClassifierLearner
+
+
+@pytest.mark.parametrize("eps", [0.001, 0.01, 0.1])
+@pytest.mark.parametrize("Constraints", [DemographicParity, EqualizedOdds])
+@pytest.mark.parametrize("h_vec", [pd.Series([1.0])])
+@pytest.mark.parametrize("opt_lambda", [True, False])
+def test_lagrangian_eval(eps, Constraints, h_vec, opt_lambda):
+    X, y, A = _get_data(A_two_dim=False)
+    estimator = LeastSquaresBinaryClassifierLearner()
+    constraints = Constraints()
+
+    # epsilon (and thereby also B) only affects L_high and L
+    B = 1 / eps
+
+    lagrangian = _Lagrangian(X, A, y, estimator, constraints, eps, B, opt_lambda=opt_lambda)
+
+    # set up initial lambda vector based on a 0-initialized theta
+    theta = pd.Series(0, lagrangian.constraints.index)
+    lambda_vec = np.exp(theta) / (1 + np.exp(theta).sum())
+
+    # call oracle to determine error and gamma and calculate exp
+    fitted_estimator = lagrangian._call_oracle(lambda_vec)
+    def h(X): return fitted_estimator.predict(X)
+    best_h_error = lagrangian.obj.gamma(h)[0]
+    best_h_gamma = lagrangian.constraints.gamma(h)
+
+    # opt_lambda affects only the calculation of L
```

Comment:
The current code tests just the internal logic of the lagrangian. Maybe that's what's intended? However two points:
* when `opt_lambda=False`, then there is a correct value of the lagrangian, so you should maybe just check that value
* when `opt_lambda=True`, then this is probably the best you can do, since the only requirement is that `projected_lambda` is an improvement (but it's left unspecified which of many possible improvements to choose)

Reply:
I'd like to explain what `opt_lambda` is for through a comment.
It seems to only decide whether or not we use `project_lambda` or just the `lambda_vec`. Does `opt_lambda` stand for "optimal lambda", or "optimize lambda", or perhaps something else? I'd like to either rename it or add a comment to explain what it is.

Reply:
We need some sort of test for the Lagrangian, otherwise we'll go into that code later, make tweaks and it'll behave differently. I'm not too happy with the tests either, but I couldn't think of anything better.

Reply:
Re. `opt_lambda`--put in a comment along the lines "`opt_lambda` indicates whether to optimize lambda during the calculation of Lagrangian". The part of the code with `opt_lambda` is going to change shortly, so don't invest too much time on that part of the code.

Reply:
Re. testing of the Lagrangian: sure--leave as is for now. I might tweak at some indefinite point in future.